files_orpg_sw/src/cpc102/tsk022/gauge_radar.1
.\" RCS info
.\" $Author: ccalvert $
.\" $Locker:  $
.\" $Date: 2011/04/13 22:53:04 $
.\" $Id: gauge_radar.1,v 1.3 2011/04/13 22:53:04 ccalvert Exp $
.\" $Revision: 1.3 $
.\" $State: Exp $
gauge_radar(1)        ORPG SUM Manpage        gauge_radar(1)

NAME:
        cpc102/tsk022 - Gauge-Radar Comparison Tool

DESCRIPTION:
        The Gauge-Radar Comparison Tool, gauge_radar, compares
        gauge readings against hourly and storm total Dual Pol
        (DP) and PPS precipitation accumulations. At present it
        compares only KOUN accumulations against surrounding OK
        gauges.

        For Dual Pol accumulations, the tool reads the
        DP_LT_ACCUM (411) buffer. The One_Hr_unbiased grid is
        used for hourly DP and the (possibly biased) Storm_Total
        grid is used for the storm DP.

        For PPS accumulations, the tool reads the HYADJSCN
        (104) buffer. The (unbiased) AccumHrly grid is used for
        the hourly PPS and the (unbiased) AccumScan grid is
        collected over the duration of the storm for storm PPS.

        For gauge accumulations, a Python script,
        gauge_radar.py, collects the gauges in KOUN_gauges.
        Each gauge in KOUN_gauges has a name, latitude, and
        longitude. The gauges under the KOUN radar umbrella are
        from the OK Mesonet, the Little Washita Watershed,
        Fort Cobb, and the OKC Micronet. To comment out a gauge,
        start the line with a '#'. An Internet connection is
        required to fetch the gauge readings. The KOUN gauges are
        not updated in real time, so gauge_radar works best as
        a playback tool.

        gauge_radar.py uses seven Python packages. To see if
        you have the Python and the packages installed, run the
        commands:

            rpm -q python

            rpm -q lapack3
            rpm -q python-dateutil
            rpm -q python-matplotlib
            rpm -q python-numpy
            rpm -q python-scipy
            rpm -q python-tz
            rpm -q refblas3

        If they are installed, the rpm should return something
        like:

            python-2.4.3-27.el5

            lapack3-3.0-19.1
            python-dateutil-1.2-1.el5.rf
            python-matplotlib-0.91.2-1.el5.rf
            python-numpy-1.2.0-1.1
            python-scipy-0.6.0-2.1
            python-tz-2006p-1.el5.rf
            refblas3-3.0-11.1

        A scratch dir, $WORK_DIR/gauges, is required by the
        Python script to return its output. $WORK_DIR should
        already be set in your $HOME/orpg_env_cshrc as
        $HOME/tmp.

        The task is volume timed. Hourly comparisons are done at
        the top of every hour, and storm comparisons are done at
        the DP storm end. Storm comparisons may also be done at
        a fixed time or every Nth volume.

        When hourly output is triggered, an HOURLY_GAUGE (179)
        product is produced, When storm output is triggered, a
        STORM_GAUGE (180) product is produced, HOURLY_GAUGE and
        STORM_GAUGE are overlays containing Gauge Name, Gauge
        Accumulation, DP Accumulation, and PPS Accumulation.
        One of Gauge Accumulation, DP Accumulation, or PPS
        Accumulation must be non-zero to be included in a
        product. Accumulations are rounded to hundredths of
        inches, and an 0 < accumulation < 0.01 inches is
        reported as 'T’ for trace. HOURLY_GAUGE and STORM_GAUGE
        may be overlayed on other RPG products.

        To view HOURLY_GAUGE and STORM_GAUGE using cvg 9.1, add
        the following lines to your $HOME/.cvg9.1/prod_config:

           179   0 0 12 .plt  0  0 .lgd .lgd .plt .plt 0 in
           180   0 0 12 .plt  0  0 .lgd .lgd .plt .plt 0 in

        When hourly or storm output is triggered, a matching
        text file is appended to. The hourly/storm text files
        are KOUN_hourly_gauge_dp_pps.txt and
        KOUN_storm_gauge_dp_pps.txt. A text file has the format:

        From 06/13/2010 21:32:05Z to 06/14/2010 18:53:52Z Storm
                                                vol 72 total 232
        Gauge ID Latitude   Longitude  AZM    KM    Gauge  DP
                                             DP-G   PPS    PPS-G
            ADAX 34.798512 -96.669090 123.7  87.21  1.150  0.743
                                            -0.407  1.012 -0.138
            ALTU 34.587219 -99.338081 247.7 186.00  0.410  0.255
                                            -0.155  0.409 -0.001
            BIXB 35.963051 -95.866211  60.3 165.69  1.330  0.609
                                            -0.721  0.614 -0.716
        ...

        vol is the last volume modulo 80 and total is the total
        number of volumes since the beginning of the run.

        After the gauges have printed, statistics on the
        gauge-radar pairs are printed:

                                   Gauges         DP        PPS

                          mean:      2.62         1.26      1.99
            standard deviation:      2.75         1.35      2.35

                                          DP-Gauge    PPS-Gauge

                        gauge-radar pairs:    126         125
                          mean field bias:      2.07        1.33
                            additive bias:   -171.03      -81.30
                               mean error:     -1.36       -0.65
                      mean absolute error:      1.37        0.88
                   root mean square error:      2.09        1.44
                       standard deviation:      1.59        1.29
                  correlation coefficient:      0.92        0.88
        fractional                   bias:     -0.52       -0.25
        fractional root mean square error:      0.80        0.54
        fractional     standard deviation:      0.61        0.49

        The statistics are also added to the product Tabular
        Alphanumeric Block.

        Due to differences in DP/PPS reporting precip. there may
        be a different number of DP-Gauge and PPS-Gauge pairs.

        The GNU Scientific Library (GSL) is used to compute
        statistics. To see if you have GSL installed, run the
        command:

            rpm -q gsl

        It should return something like:

            gsl-1.8-1.1

        If GSL is not installed, see your system administrator.

        When hourly/storm output is triggered, a matching comma
        separated values (CSV) file with the DP-Gauge and
        PPS-Gauge is appended to. The hourly/storm CSV files are
        KOUN_hourly_gauge_dp_pps.csv and
        KOUN_storm_gauge_dp_pps.csv. A CSV file has the format:

        Gauge ID,Latitude,Longitude,AZM,KM,
                                 Gauge,DP,DP-Gauge,PPS,PPS-Gauge
        MEDF,36.792419,-97.745773,351.7,175.26,
                                 0.060,0.022,-0.038,0.000,-0.060
        FREE,36.725620,-99.142342,318.1,224.70,
                                 0.010,0.102,0.092,0.252,0.242
        CHER,36.748131,-98.362740,334.5,187.00,
                                 0.040,0.082,0.042,0.130,0.090
        ALV2,36.708229,-98.709740,325.9,198.89,
                                 0.650,0.563,-0.087,0.874,0.224
        ...

        The tool takes four arguments, which are passed through
        $HOME/cfg/dea/gauge_radar.alg. The arguments control
        product timing and output:

        1. use_only_nonzero

           If 'Yes', only DP-gauge and PPS-gauge pairs where at
           least one of the gauge, DP, or PPS is non-zero will
           be added to the products, the hourly .txt file and
           the storm .txt file. If 'No', add all pairs, even
           when the gauge, DP, and PPS are all zero. The default
           is 'Yes'.

        2. stats_everywhere

           'Yes' adds the statistics set to the product TAB, the
           hourly .txt file and the storm .txt file. 'No' prints
           statistics only to the product TAB. The default is
           'Yes'.

        3. storm_print_volume

           A storm product will be generated every
           storm_print_volume. The default is to generate a
           storm product only after the end of a storm. The
           default is 10, so a storm product will be generated
           every 10th volume.

        4. storm_print_time

           The first volume after the storm_print_time produces
           a storm product. The default is to generate a storm
           product only after the end of a storm, which may be
           hard to determine. There is no default
           storm_print_time.

           Format: CCYY_MM_DD_HH_MM_SS (Zulu).

        A features overlay, with the same product id as STORM_GAUGE (180)
        has been added. It is generated every (N+1)st volume, where N
        is the storm print volume. A feature is a location plus
        optional text, like another radar or a windfarm. The features
        are set in gauge_radar_location.c. The default features are the
        KOUN gauges, without text.

TASK ID:
       10222

DATA IN:
       DP_LT_ACCUM(411)
       HYADJSCN(104)

DATA OUT:
       HOURLY_GAUGE(179)
       STORM_GAUGE(180)

SEE ALSO:

$Revision: 1.3 $


files_orpg_sw/src/cpc102/tsk022/gauge_radar.py
#!/usr/bin/python

# RCS info
# $Author: steves $
# $Locker:  $
# $Date: 2011/05/09 19:59:28 $
# $Id: gauge_radar.py,v 1.4 2011/05/09 19:59:28 steves Exp $
# $Revision: 1.4 $
# $State: Exp $


import sys, urllib2,os,cgi,time, glob

import numpy as np

# note change in "bfiedler" needed below, task defined at end

#########################################

import matplotlib  #just for the purpose of the following statement:

matplotlib.use('Agg') #prevents matplotlib from preparing to open a GUI

###########################################################

from pylab import *

import matplotlib.numerix.ma as M

from matplotlib.dates import YearLocator, MonthLocator, DayLocator, DateFormatter, HourLocator

import datetime

# jfw import pdb

####################################

mesonet_file="http://www.mesonet.org/data/public/mesonet/mdf/YYYY/MM/DD/YYYYMMDDTTTT.mdf"

#######################################################

# jfw software switch for selecting 1 of the 4 rain gauge networks
# Set a switch to 0 to not query that network

use_mesonet  = 1
use_ltwash   = 1
use_ftcobb   = 1
use_okcmicro = 1

# process parameters passed to this script:
# jfw - send it command line arguments
# jfw - removed all prints to make it silent
# jfw - made outdir 1st arguement

# print "Gathering Precipitation Data..."

outdir = sys.argv[1]

# print "Start date (yyyymmdd)?"
# start_yyyymmdd = raw_input()

start_yyyymmdd = sys.argv[2]

# print "Start time (hhmm)?"
# start_time = raw_input()

start_time = sys.argv[3]

# print "End date (yyyymmdd)?"
# end_yyyymmdd = raw_input()

end_yyyymmdd = sys.argv[4]

# print "End time (hhmm)?"
# end_time = raw_input()

end_time = sys.argv[5]

#try:

#	start_yyyymmdd = "20090903"

#	start_time = "0000"

#	end_yyyymmdd = "20090903"

#	end_time = "2000"
#except:

#	yyyymmdd="20051008" #default date

#	start_time="0000"
#	end_time = "2355"


#####################################################

# construct mesonet file name, and retrieve file

if start_time[3] != "0":

	if start_time[3] != "5":

		# print "Invalid start time!"

		sys.exit()

if end_time[3] != "0":

	if end_time[3] != "5":

		# print "Invalid end time!"

		sys.exit()


yyyy=start_yyyymmdd[0:4]

mm=start_yyyymmdd[4:6]

dd=start_yyyymmdd[6:8]

f_yyyy=end_yyyymmdd[0:4]

f_mm=end_yyyymmdd[4:6]

f_dd=end_yyyymmdd[6:8]



startingdate = datetime.date(int(yyyy),int(mm),int(dd))

endingdate = datetime.date(int(f_yyyy),int(f_mm),int(f_dd))

start_dnum = date2num(startingdate)

end_dnum = date2num(endingdate)



days_between  = int(end_dnum - start_dnum)

#print "Days Between: ", days_between


dates=[]


current_date = num2date(start_dnum)

current_dnum = start_dnum
#print "Current date: ", current_date



def read_infile(filetype, yyyy, mm, dd, end_time):

	if filetype == "meso":

		file="http://www.mesonet.org/data/public/mesonet/mdf/YYYY/MM/DD/YYYYMMDDTTTT.mdf"

	elif filetype == "ltwash":

		file="http://www.mesonet.org/data/public/ars/mdf/YYYY/MM/DD/YYYYMMDDTTTT.mdf"

	elif filetype == "ftcobb":

		file="http://www.mesonet.org/data/public/fcars/mdf/YYYY/MM/DD/YYYYMMDDTTTT.mdf"

	elif filetype == "okcmicro":

		file="http://www.mesonet.org/data/public/okc/mdf/YYYY/MM/DD/YYYYMMDDTTTT.mdf"

	else:

		print "Unrecognized file type!"		

	file=file.replace('YYYY',yyyy)
	file=file.replace('MM',mm)

	file=file.replace('DD',dd)
	file=file.replace('TTTT',end_time)

	#print 'getting file: ',file

	rains=[]
	stids=[]

	try:

		content=urllib2.urlopen(file).readlines() # this reads the text at the URL "file"

	except:

		print "Cannot open ", file

                # On 6/14/2010 22 Z, the ARS Mesonet went down due to a lightning strike.
                # We now exit, returning no gauges, but instead we should continue on
                # and get the working gauges. For future enhancement ...

		sys.exit()

                # For future enhancement ...
                # return stids, rains

	##############################################

	# process the data in the file

	nl=0

	nummissing=0

	for line in content:

                # jfw debug: print line

		nl+=1

		if nl==1 : continue #copyright line, ignore

		if nl==2 : # time header


			try:

				line=line.strip()

				year,month,day=line.split()[1:4]

				yr=int(year)

				mn=int(month)

				dy=int(day)

			except:

				# print "choked on parsing time time info"

				sys.exit()

			continue

		if nl==3 : continue # column labels, ignore

		line=line.strip()

		items=line.split()

		try:

			if filetype == "meso":

				stid,stnm,minutes,relh,tair,wspd,wvec,wdir,wdsd,wssd,wmax,rain,pres=items[0:13]

			elif filetype == "okcmicro":

				stid,stnm,time,tair,relh,pres,wspd,wvec,wdir,wdsd,wssd,wmax,rain=items[0:13]

			else:

				stid,stnm,time,rain=items[0:4]

		except:

			# print "It crashed! The line did not split into items correctly!"

			sys.exit()

		

		stids.append(stid)

		rainfloat=float(rain)

		if rainfloat < -990.:

			nummissing+=1

		# jfw DON"T Convert to inches - loses precision

		# rainfloat = rainfloat * 0.03937007

		rains.append(rainfloat)

	if end_time == "0000":

		for n in range(len(rains)):

			if rains[n] != 0.0:

				rains[n] = 0.0

			


	# print "number of missing values =",nummissing

	return stids, rains


for n in range(days_between + 1):

	# print n

	if n == 0:

		beginning = start_time

	else:

		beginning = "0000"

	if n == days_between:

		ending = end_time

	else:

		ending = "2355"

	year = str(num2date(current_dnum))[0:4]

	month = str(num2date(current_dnum))[5:7]

	day = str(num2date(current_dnum))[8:10]


        if use_mesonet == 1:

	   start_mesonet_precips = read_infile("meso",year,month,day,beginning)[1]

	   end_mesonet_precips = read_infile("meso",year,month,day,ending)[1]

        if use_ltwash == 1:

    	   start_ltwash_precips = read_infile("ltwash",year,month,day,beginning)[1]

	   end_ltwash_precips = read_infile("ltwash",year,month,day,ending)[1]

        if use_ftcobb == 1:

	   start_ftcobb_precips = read_infile("ftcobb",year,month,day,beginning)[1]

	   end_ftcobb_precips = read_infile("ftcobb",year,month,day,ending)[1]

        if use_okcmicro == 1:

	   start_okcmicro_precips = read_infile("okcmicro",year,month,day,beginning)[1]

	   end_okcmicro_precips = read_infile("okcmicro",year,month,day,ending)[1]

	if n == 0:

          if use_mesonet == 1:

		mesonet_precips = np.subtract(end_mesonet_precips, start_mesonet_precips)	

		mesonet_ids = read_infile("meso",year,month,day,beginning)[0]

          if use_ltwash == 1:

		ltwash_precips = np.subtract(end_ltwash_precips, start_ltwash_precips)

		ltwash_ids = read_infile("ltwash",year,month,day,beginning)[0]

          if use_ftcobb == 1:

		ftcobb_precips = np.subtract(end_ftcobb_precips, start_ftcobb_precips)

		ftcobb_ids = read_infile("ftcobb",year,month,day,beginning)[0]

          if use_okcmicro == 1:

		okcmicro_precips = np.subtract(end_okcmicro_precips, start_okcmicro_precips)

		okcmicro_ids = read_infile("okcmicro",year,month,day,beginning)[0]

	else:

          if use_mesonet == 1:

		mesonet_precips = np.add(mesonet_precips, np.subtract(end_mesonet_precips, start_mesonet_precips))

          if use_ltwash == 1:

		ltwash_precips = np.add(ltwash_precips, np.subtract(end_ltwash_precips, start_ltwash_precips))

          if use_ftcobb == 1:

		ftcobb_precips = np.add(ftcobb_precips, np.subtract(end_ftcobb_precips, start_ftcobb_precips))

          if use_okcmicro == 1:

		okcmicro_precips = np.add(okcmicro_precips, np.subtract(end_okcmicro_precips, start_okcmicro_precips))

	current_dnum = current_dnum + 1

# for n in range(len(mesonet_ids)):

	# print mesonet_ids[n], "%5.2f" % mesonet_precips[n]

# for n in range(len(ltwash_ids)):

	# print ltwash_ids[n], "%5.2f" % ltwash_precips[n]

# for n in range(len(ftcobb_ids)):

	# print ftcobb_ids[n], "%5.2f" % ftcobb_precips[n]

# for n in range(len(okcmicro_ids)):

	# print okcmicro_ids[n], "%5.2f" % okcmicro_precips[n]

# jfw pdb.set_trace()

# jfw added outdir

outfile = open('%s/%s%s_to_%s%s_precips.csv' % \
              (outdir, start_yyyymmdd, start_time, end_yyyymmdd, end_time), 'w')

# jfw removed some output lines

# print >>outfile, '%s, %s' % (start_yyyymmdd, end_yyyymmdd)

# print >>outfile, '%s, %s' % (start_time, end_time)

# print >>outfile, 'SiteID, Precip(in.)'

# jfw added format for mesonet_ids for easier pickup on other side.

if use_mesonet == 1:

   for n in range(len(mesonet_ids)):

	# print >>outfile, mesonet_ids[n], ", %5.2f" % mesonet_precips[n]

	print >>outfile, "%8s" % mesonet_ids[n], ", %5.2f" % mesonet_precips[n]

if use_ltwash == 1:

   for n in range(len(ltwash_ids)):

	# print >>outfile, ltwash_ids[n], ", %5.2f" % ltwash_precips[n]

	print >>outfile, "%8s" % ltwash_ids[n], ", %5.2f" % ltwash_precips[n]

if use_ftcobb == 1:

   for n in range(len(ftcobb_ids)):

   	# print >>outfile, ftcobb_ids[n], ", %5.2f" % ftcobb_precips[n]

	print >>outfile, "%8s" % ftcobb_ids[n], ", %5.2f" % ftcobb_precips[n]

if use_okcmicro == 1:

   for n in range(len(okcmicro_ids)):

	# print >>outfile, okcmicro_ids[n], ", %5.2f" % okcmicro_precips[n]

	print >>outfile, "%8s" % okcmicro_ids[n], ", %5.2f" % okcmicro_precips[n]

outfile.close()
